mobile/src/main/java/ch/epfl/sdp/mobile/state/StatefulPuzzleSelectionScreen.kt

```diff
@@ -0,0 +1,130 @@
+  )
+}
+
+private class SnapshotPuzzleSelectionScreen(
+    onPuzzleClickAction: State<(puzzle: PuzzleInfo) -> Unit>,
+    user: AuthenticatedUser,
+    facade: ChessFacade,
+) : PuzzleSelectionScreenState<PuzzleInfoAdapter> {
+
+  val onPuzzleClickAction by onPuzzleClickAction
+
+  override fun onPuzzleItemClick(puzzle: PuzzleInfoAdapter) {
+    onPuzzleClickAction(puzzle)
+  }
+
+  override val puzzles =
+      facade.unsolvedPuzzles(user).map { it.toPuzzleInfoAdapter() }.sortedBy { it.elo }
+}
+
+/**
+ * Represents the basic info of a [Puzzle] to display it in a list
+ *
+ * @param uid The [Puzzle]'s uid
+ * @param playerColor The [Color] of the player in the [Puzzle]
+ * @param elo The elo/rank (difficulty) of the puzzle
+ * @param icon The icon [Composable] to display next to the [Puzzle] description
+ */
+data class PuzzleInfoAdapter(
+    override val uid: String,
+    override val playerColor: ChessBoardState.Color,
+    override val elo: Int,
+    override val icon: @Composable () -> Unit
+) : PuzzleInfo
+
+/** Transforms a [Puzzle] to a corresponding [PuzzleInfoAdapter] */
+fun Puzzle.toPuzzleInfoAdapter(): PuzzleInfoAdapter {
+  val playerColor =
+      when (this.boardSnapshot.playing) {
+        Color.White -> ChessBoardState.Color.White
+        Color.Black -> ChessBoardState.Color.Black
+      }
+
+  val board = this.boardSnapshot.board
+  val firstMove = this.puzzleMoves.firstOrNull()?.from
+
+  val firstPiece = firstMove?.let { board[firstMove] }
+
+  val icon: @Composable () -> Unit = {
+    val desc = LocalLocalizedStrings.current.boardPieceContentDescription
+    val color = firstPiece?.color.toString()
+    val rank = firstPiece?.rank.toString()
+    when (firstPiece?.color) {
+      Color.Black -> {
+        when (firstPiece.rank) {
+          Rank.King -> Icon(ChessIcons.BlackKing, contentDescription = desc(color, rank))
+          Rank.Queen -> Icon(ChessIcons.BlackQueen, contentDescription = desc(color, rank))
+          Rank.Rook -> Icon(ChessIcons.BlackRook, contentDescription = desc(color, rank))
+          Rank.Bishop -> Icon(ChessIcons.BlackBishop, contentDescription = desc(color, rank))
+          Rank.Knight -> Icon(ChessIcons.BlackKnight, contentDescription = desc(color, rank))
+          Rank.Pawn -> Icon(ChessIcons.BlackPawn, contentDescription = desc(color, rank))
+        }
+      }
+      Color.White -> {
+        when (firstPiece.rank) {
+          Rank.King -> Icon(ChessIcons.WhiteKing, contentDescription = desc(color, rank))
+          Rank.Queen -> Icon(ChessIcons.WhiteQueen, contentDescription = desc(color, rank))
+          Rank.Rook -> Icon(ChessIcons.WhiteRook, contentDescription = desc(color, rank))
+          Rank.Bishop -> Icon(ChessIcons.WhiteBishop, contentDescription = desc(color, rank))
+          Rank.Knight -> Icon(ChessIcons.WhiteKnight, contentDescription = desc(color, rank))
+          Rank.Pawn -> Icon(ChessIcons.WhitePawn, contentDescription = desc(color, rank))
+        }
```

Comment:
Question : could you use this function from `ClassicChessBoard.kt` :

<img width="663" alt="image" src="https://user-images.githubusercontent.com/6318990/167519354-9eb37247-7ed8-4cae-a985-4a5f2d8c703d.png">


alongside with the `MatchChessBoardState.Companion` utility functions to avoid the duplicate mapping to the icons ?

Reply:
Well yes, but actually no. I need the content description as well. This could come in a subsequent refactor.